// prettier-plugin-solidity/dist/slang-nodes/ArgumentsDeclaration.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { PositionalArgumentsDeclaration } from './PositionalArgumentsDeclaration.js';
import { NamedArgumentsDeclaration } from './NamedArgumentsDeclaration.js';
export class ArgumentsDeclaration {
    constructor(ast, options) {
        this.kind = NonterminalKind.ArgumentsDeclaration;
        let metadata = getNodeMetadata(ast);
        switch (ast.variant.cst.kind) {
            case NonterminalKind.PositionalArgumentsDeclaration:
                this.variant = new PositionalArgumentsDeclaration(ast.variant, options);
                break;
            case NonterminalKind.NamedArgumentsDeclaration:
                this.variant = new NamedArgumentsDeclaration(ast.variant, options);
                break;
            default:
                throw new Error(`Unexpected variant: ${ast.variant.cst.kind}`);
        }
        metadata = updateMetadata(metadata, [this.variant]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        return path.call(print, 'variant');
    }
}
