// prettier-plugin-solidity/dist/slang-nodes/ConstructorDefinition.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { printFunction } from '../slang-printers/print-function.js';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { ParametersDeclaration } from './ParametersDeclaration.js';
import { ConstructorAttributes } from './ConstructorAttributes.js';
import { Block } from './Block.js';
export class ConstructorDefinition {
    constructor(ast, options) {
        this.kind = NonterminalKind.ConstructorDefinition;
        let metadata = getNodeMetadata(ast);
        this.parameters = new ParametersDeclaration(ast.parameters, options);
        this.attributes = new ConstructorAttributes(ast.attributes, options);
        this.body = new Block(ast.body, options);
        metadata = updateMetadata(metadata, [
            this.parameters,
            this.attributes,
            this.body
        ]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        return printFunction('constructor', this, path, print);
    }
}
