// prettier-plugin-solidity/dist/slang-nodes/DoWhileStatement.js
import { doc } from 'prettier';
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { printSeparatedItem } from '../slang-printers/print-separated-item.js';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { Statement } from './Statement.js';
import { Expression } from './Expression.js';
const { group, indent, line } = doc.builders;
export class DoWhileStatement {
    constructor(ast, options) {
        this.kind = NonterminalKind.DoWhileStatement;
        let metadata = getNodeMetadata(ast);
        this.body = new Statement(ast.body, options);
        this.condition = new Expression(ast.condition, options);
        metadata = updateMetadata(metadata, [this.body, this.condition]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        return [
            'do',
            this.body.variant.kind === NonterminalKind.Block
                ? [' ', path.call(print, 'body'), ' ']
                : group([indent([line, path.call(print, 'body')]), line]),
            'while (',
            printSeparatedItem(path.call(print, 'condition')),
            ');'
        ];
    }
}
