// prettier-plugin-solidity/dist/slang-nodes/FallbackFunctionDefinition.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { printFunction } from '../slang-printers/print-function.js';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { ParametersDeclaration } from './ParametersDeclaration.js';
import { FallbackFunctionAttributes } from './FallbackFunctionAttributes.js';
import { ReturnsDeclaration } from './ReturnsDeclaration.js';
import { FunctionBody } from './FunctionBody.js';
export class FallbackFunctionDefinition {
    constructor(ast, options) {
        this.kind = NonterminalKind.FallbackFunctionDefinition;
        let metadata = getNodeMetadata(ast);
        this.parameters = new ParametersDeclaration(ast.parameters, options);
        this.attributes = new FallbackFunctionAttributes(ast.attributes, options);
        if (ast.returns) {
            this.returns = new ReturnsDeclaration(ast.returns, options);
        }
        this.body = new FunctionBody(ast.body, options);
        metadata = updateMetadata(metadata, [
            this.parameters,
            this.attributes,
            this.returns,
            this.body
        ]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
        this.cleanModifierInvocationArguments();
    }
    cleanModifierInvocationArguments() {
        for (const attribute of this.attributes.items) {
            if (typeof attribute.variant !== 'string' &&
                attribute.variant.kind === NonterminalKind.ModifierInvocation) {
                attribute.variant.cleanModifierInvocationArguments();
            }
        }
    }
    print(path, print) {
        return printFunction('fallback', this, path, print);
    }
}
