// prettier-plugin-solidity/dist/slang-nodes/ForStatementInitialization.js
import { NonterminalKind, TerminalNode } from '@nomicfoundation/slang/cst';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { ExpressionStatement } from './ExpressionStatement.js';
import { VariableDeclarationStatement } from './VariableDeclarationStatement.js';
import { TupleDeconstructionStatement } from './TupleDeconstructionStatement.js';
export class ForStatementInitialization {
    constructor(ast, options) {
        this.kind = NonterminalKind.ForStatementInitialization;
        let metadata = getNodeMetadata(ast);
        if (ast.variant instanceof TerminalNode) {
            this.variant = ast.variant.unparse();
        }
        else {
            switch (ast.variant.cst.kind) {
                case NonterminalKind.ExpressionStatement:
                    this.variant = new ExpressionStatement(ast.variant, options);
                    break;
                case NonterminalKind.VariableDeclarationStatement:
                    this.variant = new VariableDeclarationStatement(ast.variant, options);
                    break;
                case NonterminalKind.TupleDeconstructionStatement:
                    this.variant = new TupleDeconstructionStatement(ast.variant, options);
                    break;
                default:
                    throw new Error(`Unexpected variant: ${ast.variant.cst.kind}`);
            }
        }
        metadata = updateMetadata(metadata, typeof this.variant === 'string' ? [] : [this.variant]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        return typeof this.variant === 'string'
            ? this.variant
            : path.call(print, 'variant');
    }
}
