// prettier-plugin-solidity/dist/slang-nodes/ImportClause.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { PathImport } from './PathImport.js';
import { NamedImport } from './NamedImport.js';
import { ImportDeconstruction } from './ImportDeconstruction.js';
export class ImportClause {
    constructor(ast, options) {
        this.kind = NonterminalKind.ImportClause;
        let metadata = getNodeMetadata(ast);
        switch (ast.variant.cst.kind) {
            case NonterminalKind.PathImport:
                this.variant = new PathImport(ast.variant, options);
                break;
            case NonterminalKind.NamedImport:
                this.variant = new NamedImport(ast.variant, options);
                break;
            case NonterminalKind.ImportDeconstruction:
                this.variant = new ImportDeconstruction(ast.variant, options);
                break;
            default:
                throw new Error(`Unexpected variant: ${ast.variant.cst.kind}`);
        }
        metadata = updateMetadata(metadata, [this.variant]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        return path.call(print, 'variant');
    }
}
