// prettier-plugin-solidity/dist/slang-nodes/ModifierInvocation.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { IdentifierPath } from './IdentifierPath.js';
import { ArgumentsDeclaration } from './ArgumentsDeclaration.js';
export class ModifierInvocation {
    constructor(ast, options) {
        this.kind = NonterminalKind.ModifierInvocation;
        let metadata = getNodeMetadata(ast);
        this.name = new IdentifierPath(ast.name);
        if (ast.arguments) {
            this.arguments = new ArgumentsDeclaration(ast.arguments, options);
        }
        metadata = updateMetadata(metadata, [this.name, this.arguments]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    cleanModifierInvocationArguments() {
        if (this.arguments &&
            this.arguments.variant.kind ===
                NonterminalKind.PositionalArgumentsDeclaration &&
            this.arguments.variant.isEmpty()) {
            delete this.arguments;
        }
    }
    print(path, print) {
        return [path.call(print, 'name'), path.call(print, 'arguments')];
    }
}
