// prettier-plugin-solidity/dist/slang-nodes/OrExpression.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { printLogicalOperation } from '../slang-printers/print-logical-operation.js';
import { createHugFunction } from '../slang-utils/create-hug-function.js';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { Expression } from './Expression.js';
const tryToHug = createHugFunction(['&&']);
export class OrExpression {
    constructor(ast, options) {
        this.kind = NonterminalKind.OrExpression;
        let metadata = getNodeMetadata(ast);
        this.leftOperand = new Expression(ast.leftOperand, options);
        this.operator = ast.operator.unparse();
        this.rightOperand = new Expression(ast.rightOperand, options);
        metadata = updateMetadata(metadata, [this.leftOperand, this.rightOperand]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
        this.leftOperand = tryToHug(this.leftOperand);
        this.rightOperand = tryToHug(this.rightOperand);
    }
    print(path, print, options) {
        return printLogicalOperation(this, path, print, options);
    }
}
