// prettier-plugin-solidity/dist/slang-nodes/PositionalArguments.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { printComments } from '../slang-printers/print-comments.js';
import { printSeparatedItem } from '../slang-printers/print-separated-item.js';
import { printSeparatedList } from '../slang-printers/print-separated-list.js';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { Expression } from './Expression.js';
export class PositionalArguments {
    constructor(ast, options) {
        this.kind = NonterminalKind.PositionalArguments;
        let metadata = getNodeMetadata(ast, true);
        this.items = ast.items.map((item) => new Expression(item, options));
        this.separators = ast.separators.map((separator) => separator.unparse());
        metadata = updateMetadata(metadata, [this.items]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        if (this.items.length > 0) {
            return printSeparatedList(path.map(print, 'items'));
        }
        const argumentComments = printComments(path);
        return argumentComments.length > 0
            ? printSeparatedItem(argumentComments)
            : '';
    }
}
