// prettier-plugin-solidity/dist/slang-nodes/Pragma.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { AbicoderPragma } from './AbicoderPragma.js';
import { ExperimentalPragma } from './ExperimentalPragma.js';
import { VersionPragma } from './VersionPragma.js';
export class Pragma {
    constructor(ast, options) {
        this.kind = NonterminalKind.Pragma;
        let metadata = getNodeMetadata(ast);
        switch (ast.variant.cst.kind) {
            case NonterminalKind.AbicoderPragma:
                this.variant = new AbicoderPragma(ast.variant);
                break;
            case NonterminalKind.ExperimentalPragma:
                this.variant = new ExperimentalPragma(ast.variant, options);
                break;
            case NonterminalKind.VersionPragma:
                this.variant = new VersionPragma(ast.variant);
                break;
            default:
                throw new Error(`Unexpected variant: ${ast.variant.cst.kind}`);
        }
        metadata = updateMetadata(metadata, [this.variant]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        return path.call(print, 'variant');
    }
}
