// prettier-plugin-solidity/dist/slang-nodes/TupleMember.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { TypedTupleMember } from './TypedTupleMember.js';
import { UntypedTupleMember } from './UntypedTupleMember.js';
export class TupleMember {
    constructor(ast, options) {
        this.kind = NonterminalKind.TupleMember;
        let metadata = getNodeMetadata(ast);
        switch (ast.variant.cst.kind) {
            case NonterminalKind.TypedTupleMember:
                this.variant = new TypedTupleMember(ast.variant, options);
                break;
            case NonterminalKind.UntypedTupleMember:
                this.variant = new UntypedTupleMember(ast.variant);
                break;
            default:
                throw new Error(`Unexpected variant: ${ast.variant.cst.kind}`);
        }
        metadata = updateMetadata(metadata, [this.variant]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        return path.call(print, 'variant');
    }
}
