// prettier-plugin-solidity/dist/slang-nodes/TypeName.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { ArrayTypeName } from './ArrayTypeName.js';
import { FunctionType } from './FunctionType.js';
import { MappingType } from './MappingType.js';
import { ElementaryType } from './ElementaryType.js';
import { IdentifierPath } from './IdentifierPath.js';
export class TypeName {
    constructor(ast, options) {
        this.kind = NonterminalKind.TypeName;
        let metadata = getNodeMetadata(ast);
        switch (ast.variant.cst.kind) {
            case NonterminalKind.ArrayTypeName:
                this.variant = new ArrayTypeName(ast.variant, options);
                break;
            case NonterminalKind.FunctionType:
                this.variant = new FunctionType(ast.variant, options);
                break;
            case NonterminalKind.MappingType:
                this.variant = new MappingType(ast.variant, options);
                break;
            case NonterminalKind.ElementaryType:
                this.variant = new ElementaryType(ast.variant);
                break;
            case NonterminalKind.IdentifierPath:
                this.variant = new IdentifierPath(ast.variant);
                break;
            default:
                throw new Error(`Unexpected variant: ${ast.variant.cst.kind}`);
        }
        metadata = updateMetadata(metadata, [this.variant]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        return path.call(print, 'variant');
    }
}
