// prettier-plugin-solidity/dist/slang-nodes/VersionExpression.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { VersionRange } from './VersionRange.js';
import { VersionTerm } from './VersionTerm.js';
export class VersionExpression {
    constructor(ast) {
        this.kind = NonterminalKind.VersionExpression;
        let metadata = getNodeMetadata(ast);
        switch (ast.variant.cst.kind) {
            case NonterminalKind.VersionRange:
                this.variant = new VersionRange(ast.variant);
                break;
            case NonterminalKind.VersionTerm:
                this.variant = new VersionTerm(ast.variant);
                break;
            default:
                throw new Error(`Unexpected variant: ${ast.variant.cst.kind}`);
        }
        metadata = updateMetadata(metadata, [this.variant]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        return path.call(print, 'variant');
    }
}
