// prettier-plugin-solidity/dist/slang-nodes/YulExpression.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { YulFunctionCallExpression } from './YulFunctionCallExpression.js';
import { YulLiteral } from './YulLiteral.js';
import { YulPath } from './YulPath.js';
export class YulExpression {
    constructor(ast, options) {
        this.kind = NonterminalKind.YulExpression;
        let metadata = getNodeMetadata(ast);
        switch (ast.variant.cst.kind) {
            case NonterminalKind.YulFunctionCallExpression:
                this.variant = new YulFunctionCallExpression(ast.variant, options);
                break;
            case NonterminalKind.YulLiteral:
                this.variant = new YulLiteral(ast.variant, options);
                break;
            case NonterminalKind.YulPath:
                this.variant = new YulPath(ast.variant);
                break;
            default:
                throw new Error(`Unexpected variant: ${ast.variant.cst.kind}`);
        }
        metadata = updateMetadata(metadata, [this.variant]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        return path.call(print, 'variant');
    }
}
