// prettier-plugin-solidity/dist/slang-nodes/YulLiteral.js
import { NonterminalKind, TerminalNode } from '@nomicfoundation/slang/cst';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { HexStringLiteral } from './HexStringLiteral.js';
import { StringLiteral } from './StringLiteral.js';
export class YulLiteral {
    constructor(ast, options) {
        this.kind = NonterminalKind.YulLiteral;
        let metadata = getNodeMetadata(ast);
        if (ast.variant instanceof TerminalNode) {
            this.variant = ast.variant.unparse();
        }
        else {
            switch (ast.variant.cst.kind) {
                case NonterminalKind.HexStringLiteral:
                    this.variant = new HexStringLiteral(ast.variant, options);
                    break;
                case NonterminalKind.StringLiteral:
                    this.variant = new StringLiteral(ast.variant, options);
                    break;
                default:
                    throw new Error(`Unexpected variant: ${ast.variant.cst.kind}`);
            }
        }
        metadata = updateMetadata(metadata, typeof this.variant === 'string' ? [] : [this.variant]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        return typeof this.variant === 'string'
            ? this.variant
            : path.call(print, 'variant');
    }
}
