// prettier-plugin-solidity/dist/slang-nodes/YulSwitchCase.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { YulDefaultCase } from './YulDefaultCase.js';
import { YulValueCase } from './YulValueCase.js';
export class YulSwitchCase {
    constructor(ast, options) {
        this.kind = NonterminalKind.YulSwitchCase;
        let metadata = getNodeMetadata(ast);
        switch (ast.variant.cst.kind) {
            case NonterminalKind.YulDefaultCase:
                this.variant = new YulDefaultCase(ast.variant, options);
                break;
            case NonterminalKind.YulValueCase:
                this.variant = new YulValueCase(ast.variant, options);
                break;
            default:
                throw new Error(`Unexpected variant: ${ast.variant.cst.kind}`);
        }
        metadata = updateMetadata(metadata, [this.variant]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
    }
    print(path, print) {
        return path.call(print, 'variant');
    }
}
